### PR TITLE
Add `Project` information to `Cluster` resource

### PR DESCRIPTION
**Pull request by SimonKienzler**
**How to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:

The `extensionsv1alpha1.Cluster` type is already very useful for obtaining relevant data concerning `Seed`, `Shoot` and `Infrastructure`. For our use cases, it would be really handy to also have the `Project` resource present in the `Cluster` extension. As the project data was already available where the `Cluster` resource is created, this is a rather small change - but greatly improves developer experience for us.

Background: We would like to use `Project`-specific data in an extension on seeds.

**Release note**:

```feature developer
Project data is available in the Cluster extension on seeds.
```

### CODE REVIEW

**Review by timebertt — COMMENTED** [review]
Awesome!
/lgtm

**Comment by timebertt** [comment]
https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7106/pull-gardener-integration/1597618226601136128 is another occurrence of https://github.com/gardener/gardener/issues/6933

/test pull-gardener-integration

File: pkg/extensions/cluster.go [hunk 1]
```diff
@@ -84,6 +86,15 @@ func SyncClusterResourceToSeed(
 		shootObj.ManagedFields = nil
 	}
 
+	if project != nil {
+		projectObj = project.DeepCopy()
+		projectObj.TypeMeta = metav1.TypeMeta{
+			APIVersion: gardencorev1beta1.SchemeGroupVersion.String(),
+			Kind:       "Project",
+		}
+		shootObj.ManagedFields = nil
```
Comment (vpnachev) [inline pkg/extensions/cluster.go]:
```suggestion
		projectObj.ManagedFields = nil
```

**Review by vpnachev — CHANGES_REQUESTED** [review]
> Background: We would like to use Project-specific data in an extension on seeds.


Can you tell us a bit more about your use case? I am interested to see if it is possible to be shifted from the seed to the garden cluster?

The `cluster` resource is already bloated and increasing its size will reflect on the memory and network usage, also the etcd of the seed will have to serve this additional data.
Pretty much each and every extension is reading the cluster, mostly just see if a shoot is hibernated or its k8s version, I mean they read hundreds of KB just to check two boolean fields or a several bytes long string.

**Comment by SimonKienzler** [comment]
@vpnachev

> Can you tell us a bit more about your use case?

Sure. We have custom annotations and labels on the `Project` resource which contain information that is important to have for us in order to enhance metrics coming from the shoot. We therefore would like to access these annotations and labels in the shoot control plane on our seeds, where the monitoring is configured.

**Comment by vpnachev** [comment]
Adding the entire project just for the annotations and labels looks too much to me. 
Instead, I think you can do one of the following
- implement a controller or admission webhook that syncs the project labels and annotations with the shoot one. 
- put the required information in a separate resource and let gardenlet sync it via https://github.com/gardener/gardener/blob/684d5b142c9da6b2df1982048121aa70c9ee4197/pkg/apis/core/v1beta1/types_shoot.go#L116-L118 to the seed
- think about some custom resource in the garden cluster that configures the shoot observability and let gardenlet sync it to the seed, or just re-use the `Extension`. You can take a look at https://github.com/gardener/gardener/blob/master/pkg/apis/settings/v1alpha1/types_openidconnect_preset.go and https://github.com/gardener/gardener/blob/master/pkg/apis/settings/v1alpha1/types_cluster_openidconnect_preset.go for inspiration.

**Comment by timebertt** [comment]
> The `cluster` resource is already bloated and increasing its size will reflect on the memory and network usage, also the etcd of the seed will have to serve this additional data.

I agree. Currently, we also sync the kubectl last applied annotation to the `Cluster` resource which could be filtered out if we are concerned about size in etcd and network transfer.

> Pretty much each and every extension is reading the cluster, mostly just see if a shoot is hibernated or its k8s version, I mean they read hundreds of KB just to check two boolean fields or a several bytes long string.

Please keep in mind, that extensions use cached clients and thus don't transfer the entire `Cluster` object just for checking parts of it.

> Adding the entire project just for the annotations and labels looks too much to me.
> ...
> implement a controller or admission webhook that syncs the project labels and annotations with the shoot one.

We are aware, that our current use case can also be solved by copying project information to shoot annotations/labels.
However, there are other valid use cases where extensions might want to act, for example, on the specified project members.

IMO, the `Cluster` resource has exactly this purpose: to provide additional information on the shoot and its environment (e.g., `Seed`, `CloudProfile`, and `Project`) to extension controllers that need this information.
Also, the `Project` object is typically very small in comparison to the other contained objects.
If we want to reduce this duplication because of etcd storage and network transfer, I would suggest optimizing at other places, e.g. removing the kubectl annotation, or copying the seed object just once instead of per shoot.

In short: I don't see a problem with adding the `Project` object to the `Cluster` resource, as it fits the purpose and doesn't make things much worse.

Any other opinions? @rfranzke @timuthy

**Comment by donistz** [comment]
> Also, the `Project` object is typically very small in comparison to the other contained objects.

'Project' object can be also big when consumers configure a large number of project members. Also, when you need only some tagging to be replicated in each and every cluster resource and you don't need members' data also replicated, it is better to find a way to replicate only the data necessary or at least to give the option of the hosting party to decide centrally what will be replicated and when if at all. In your use case you don't talk about members at all. Members are not known by the seeds/shoots and not required as data there. If you don't need the members, you are trying to propose a solution for your use case that is exposing a risk by replicating data to a place where it is not necessary.

**Comment by vpnachev** [comment]
> Please keep in mind, that extensions use cached clients and thus don't transfer the entire `Cluster` object just for checking parts of it.

I know about the cache, while it saves network traffic - it consumes runtime memory. 

>  However, there are other valid use cases where extensions might want to act, for example, on the specified project members.

This is wrong - not all entities involved in a Gardener project must be declared in the project spec as the project members are authorized to define their own roles and rolebindings https://github.com/gardener/gardener/blob/8917b4905d6da6396833adbaee3bd64a4dec2b6b/charts/gardener/controlplane/charts/application/templates/rbac-user.yaml#L374-L387

So having the list of the project members in the seed and doing some authorization assumptions based on it is not correct to the actual RBAC state in the garden cluster.

**Comment by timebertt** [comment]
Ok, as written above, we can also solve our specific use case by syncing project information to shoot annotations.
We wanted to take the long road and open up gardener for more general use cases that we considered useful.
However, it seems like such general use cases are not considered valid by others, so I propose going back to the short path and solving our specific use case only by the means discussed in this PR.

In case @SimonKienzler agrees, I suggest closing this PR.
